extensions/amp-story/1.0/amp-story.js

```diff
@@ -1624,8 +1633,35 @@ export class AmpStory extends AMP.BaseElement {
    * @visibleForTesting
    */
   onResize() {
-    const uiState = this.getUIType_();
-    this.storeService_.dispatch(Action.TOGGLE_UI, uiState);
+    const previousUiState = this.uiState_;
+    this.uiState_ = this.getUIType_();
+
+    const wasMobile = previousUiState === UIType.MOBILE;
+    const inputHasFocus = this.win.document.activeElement?.tagName === 'INPUT';
+    const softKeyboardIsProbablyOpen = wasMobile && inputHasFocus;
+    if (!softKeyboardIsProbablyOpen) {
+      // The opening of the Android soft keyboard triggers a viewport resize
+      // that can cause the story's dimensions to appear to be those of a
+      // desktop. Here, we assume that the soft keyboard is open if the story
+      // has resized while an input element has focus, and we do not toggle the
+      // UI.
+      this.storeService_.dispatch(Action.TOGGLE_UI, this.uiState_);
+    }
+
+    if (this.uiState_ === UIType.MOBILE) {
+      const currentHeight = this.getViewport().getHeight();
+      if (currentHeight > this.maxViewportHeight_) {
+        this.maxViewportHeight_ = currentHeight;
+        // Maintain story height on viewport resize to prevent the story page
+        // from resizing. This is particularly necessary on Android, when
+        // opening or closing the soft keyboard triggers a viewport resize.
+        this.mutateElement(() => {
+          setImportantStyles(this.element, {
+            'height': px(this.maxViewportHeight_),
+          });
+        });
+      }
+    }
```

Comment:
This block is assuming that all mobile resizes are caused by the soft keyboard. `maxViewportHeight_` could have changed between `buildCallback` and when this code runs for other reasons

Reply:
Is there any case where the mobile story shouldn't be the height of the viewport? If not, then we can treat this block like a general safeguard that prevents mobile stories from ever having an incorrect height. But if there are cases where the story and viewport heights are expected to differ, then maybe we can do the following:
  * Add a new `this.isSoftKeyboardOpen_` property to `amp-story.js`
  * Set the `height` style to the viewport size when the soft keyboard is open
  * Remove the `height` style when the soft keyboard is closed